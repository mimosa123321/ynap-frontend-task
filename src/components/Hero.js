import React from "react";
import { useEffect, useState, useRef } from "react";
import { isMobileView } from "../utils/utils";
import "../styles/styles.scss";

const Hero = (props) => {
  const { scrollPosition } = props;
  const [positionY, setPositionY] = useState(0);
  const [opacity, setOpacity] = useState(1);
  const [bgPosY, setBgPosY] = useState(1);
  const heroContainer = useRef(null);

  useEffect(() => {
    const speed = 0.2;
    const maxPosOffset = 30;
    const posY = parseFloat((scrollPosition * speed).toString());
    let opacity = 1;
    if (heroContainer.current) {
      const diff = scrollPosition - heroContainer.current.offsetTop;
      if (diff > 0) {
        opacity = 1 - diff * (1 / (heroContainer.current.offsetHeight * speed));
      }
    }
    setPositionY(posY > maxPosOffset ? maxPosOffset : posY);
    setOpacity(opacity < 0 ? 0 : opacity);
    setBgPosY(isMobileView() ? 0 : posY);
  }, [scrollPosition]);

  return (
    <div
      className="hero-container"
      style={{ backgroundPositionY: `-${bgPosY}px` }}
      ref={heroContainer}
    >
      <div className="hero-container-cover" style={{ opacity: `${opacity}` }}>
        <div
          className="contents"
          style={{ transform: `translate3d(0, ${positionY}px, 0)` }}
        >
          <div className="type">cover story</div>
          <div className="title">standing strong</div>
          <div className="author">
            <i>
              with
              <br />
              Danal Gurlen
            </i>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Hero;
