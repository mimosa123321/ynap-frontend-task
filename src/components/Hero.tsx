import { useScroll } from '../hooks/useScroll';

import "../styles/styles.scss";
import {useEffect} from "react";

const Hero = () => {
    const { scrollPosition, addScrollHandler, removeScrollHandler } = useScroll();

    useEffect( () => {
        addScrollHandler();
    }, []);

    useEffect( () => {
        // console.log(scrollPosition);
    }, [scrollPosition])

    return (
        <div className="hero-container">
            <div className="hero-container-cover">
                <div className="contents">
                    <div className="type">cover story</div>
                    <div className="title">standing strong</div>
                    <div className="author"><i>with<br/>Danal Gurlen</i></div>
                </div>
            </div>

        </div>
    );
}

export default Hero;
