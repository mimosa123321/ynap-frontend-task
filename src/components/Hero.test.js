import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Hero from "./Hero";

test("renders correctly", () => {
  const component = mount(<Hero scrollPosition={0} />);
  expect(toJson(component)).toMatchSnapshot();
});

test("the opacity of hero container cover should be 1 when scrollPosition equal to 0", () => {
  const component = mount(<Hero scrollPosition={0} />);
  const child = component.find(".hero-container-cover");
  expect(child.get(0).props.style.opacity).toStrictEqual("1");
});

test("the opacity of hero container cover should be 0 when scrollPosition equal to 1000", () => {
  const component = mount(<Hero scrollPosition={1000} />);
  const child = component.find(".hero-container-cover");
  expect(child.get(0).props.style.opacity).toStrictEqual("0");
});
