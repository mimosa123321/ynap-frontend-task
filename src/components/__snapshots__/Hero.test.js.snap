// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Hero
  scrollPosition={0}
>
  <div
    className="hero-container"
    style={
      Object {
        "backgroundPositionY": "-0px",
      }
    }
  >
    <div
      className="hero-container-cover"
      style={
        Object {
          "opacity": "1",
        }
      }
    >
      <div
        className="contents"
        style={
          Object {
            "transform": "translate3d(0, 0px, 0)",
          }
        }
      >
        <div
          className="type"
        >
          cover story
        </div>
        <div
          className="title"
        >
          standing strong
        </div>
        <div
          className="author"
        >
          <i>
            with
            <br />
            Danal Gurlen
          </i>
        </div>
      </div>
    </div>
  </div>
</Hero>
`;

exports[`the opacity of hero container cover should be 0 when scrollPosition equal to 1000 1`] = `"0"`;

exports[`the opacity of hero container cover should disappear when scrollPosition equal to 300 1`] = `
<div
  className="hero-container"
  style={
    Object {
      "backgroundPositionY": "-1px",
    }
  }
>
  <div
    className="hero-container-cover"
    style={
      Object {
        "opacity": "1",
      }
    }
  >
    <div
      className="contents"
      style={
        Object {
          "transform": "translate3d(0, 0px, 0)",
        }
      }
    >
      <div
        className="type"
      >
        cover story
      </div>
      <div
        className="title"
      >
        standing strong
      </div>
      <div
        className="author"
      >
        <i>
          with
          <br />
          Danal Gurlen
        </i>
      </div>
    </div>
  </div>
</div>
`;
